### PR TITLE
Added unit tests for feature dev planning phase

### PR DESCRIPTION
**Pull request by kumsmrit**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

- Added unit tests for feature dev planning phase

## Checklist
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    private lateinit var spySession: Session
+    private lateinit var featureDevClient: FeatureDevClient
+
+    private val tabId = "tabId"
+
+    @Before
+    override fun setup() {
+        super.setup()
+        featureDevClient = mock()
+        messanger = mock()
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:58]: typo: instead of "messanger" it should be "messenger"

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            on { project }.thenReturn(project)
+            on { messagesFromAppToUi }.thenReturn(messanger)
+        }
+        authController = spy(AuthController())
+        spySession = spy(Session("tabId", project))
+        mockkStatic("software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.UploadArtifactKt")
+        every { uploadArtifactToS3(any(), any(), any(), any(), any()) } just runs
+
+        ToolWindowManager.getInstance(projectRule.project).registerToolWindow(
+            RegisterToolWindowTask(id = AmazonQToolWindowFactory.WINDOW_ID, canWorkInDumbMode = true)
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:70]: question: why is this needed?

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Test
+    fun `test new tab opened`() {
+        val message: IncomingFeatureDevMessage.NewTabCreated = IncomingFeatureDevMessage.NewTabCreated("new-tab-created", tabId)
+        whenever(chatSessionStorage.getSession(any(), any())).thenReturn(spySession)
+
+        runTest {
+            controller.processNewTabCreatedMessage(message)
+        }
+        verify(authController, times(1)).getAuthNeededStates(any())
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:84]: suggestion: here it could be tested if the getAuthNeededStates is called with the mock project created above, rather than any input.

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        ToolWindowManager.getInstance(projectRule.project).registerToolWindow(
+            RegisterToolWindowTask(id = AmazonQToolWindowFactory.WINDOW_ID, canWorkInDumbMode = true)
+        )
+
+        controller = FeatureDevController(appContext, chatSessionStorage, authController)
+    }
+
+    @Test
+    fun `test new tab opened`() {
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:77]: suggestion: adding a non happy path for this `newTabOpened` function, where it could be tested if the `sendError` is being really called with a mocked error message.

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    @Before
+    override fun setup() {
+        featureDevClient = mock()
+        repoContext = mock()
+        sessionStateConfig = SessionStateConfig(testConversationId, featureDevClient, repoContext)
+        refinementState = RefinementState("", "tabId", sessionStateConfig, exampleCreateUploadUrlResponse.uploadId(), 0)
+    }
+
+    @Test(expected = FeatureDevException::class)
+    fun `test refinement state with no userMssg`() {
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt:48]: suggestion: could we test the message in the exception is the UX defined one?

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        assertThat(actual.nextState).isInstanceOf(RefinementState::class.java)
+        assertThat(actual.interaction.interactionSucceeded).isTrue()
+
+        verify(featureDevClient, times(1)).generateTaskAssistPlan(any(), any(), any())
+        assertThat(refinementState.phase).isEqualTo(SessionStatePhase.APPROACH)
+        assertThat(refinementState.approach).isEqualTo(exampleGenerateTaskAssistPlanResult.approach)
+    }
+
+    @Test
+    fun `test refinement state with failed approach`() = runTest {
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt:71]: suggestion: shall we test other error case scenarios from generateApproach failures?

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    private lateinit var batcher: TelemetryBatcher
+
+    private val uploadId = "test-upload-id"
+
+    @Before
+    override fun setup() {
+        featureDevClient = mock()
+        batcher = mock<TelemetryBatcher>()
+        telemetryService = TestTelemetryService(batcher = batcher)
+        ApplicationManager.getApplication().replaceService(TelemetryService::class.java, telemetryService, disposableRule.disposable)
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt:57]: question: why is the telemetry service being mocked here?

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+    @Test
+    fun `test createConversation`() {
+        whenever(featureDevClient.createTaskAssistConversation()).thenReturn(exampleCreateTaskAssistConversationResponse)
+        val actual = createConversation(featureDevClient)
+        assertThat(actual).isEqualTo(exampleCreateTaskAssistConversationResponse.conversationId())
+    }
+
+    @Test
+    fun `test createConversation with error`() {
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt:68]: nitpick: testing also throwing a CodeWhispererRuntimeException.

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        whenever(featureDevClient.createTaskAssistConversation()).thenThrow(RuntimeException())
+        assertThatThrownBy {
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt:70]: suggestion: here you could use 

```assertThrows<Exception>(message="message") {
function
}```

to have a more complete check.

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/SessionTest.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        assertThat(session.isAuthenticating).isFalse()
+    }
+
+    @Test
+    fun `test preloader`() = runTest {
+        whenever(featureDevClient.createTaskAssistConversation()).thenReturn(exampleCreateTaskAssistConversationResponse)
+
+        session.preloader(userMessage, messanger)
+        assertThat(session.conversationId).isEqualTo(testConversationId)
+        assertThat(session.sessionState).isInstanceOf(PrepareRefinementState::class.java)
```
Comment (sannicm) [inline plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/SessionTest.kt:50]: suggestion: could we test also that if we call two times preloader, the second time createTaskAssist shouldn't be called?

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt (hunk 6 above)
Comment (sannicm) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt:71]: Discussed offline: Those error scenarios are being tested in the FeatureDevClientUtil tests.

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt (hunk 1 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:58]: updated

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt (hunk 2 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:70]: not needed, have removed

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt (hunk 3 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt:84]: Updated `getAuthNeededStates` and verified for `getSessionInfo` method being called

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt (hunk 5 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/RefinementStateTest.kt:48]: Updated test to verify for message

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt (hunk 7 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/util/FeatureDevClientUtilTest.kt:57]: not needed, have removed

File: plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/SessionTest.kt (hunk 10 above)
Comment (kumsmrit) [reply plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/session/SessionTest.kt:50]: Updated test to check for createTaskAssist being called only once
